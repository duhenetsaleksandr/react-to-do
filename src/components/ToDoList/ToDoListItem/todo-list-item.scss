@use '../../../styles/colors';

.todos__list__item {
    display: flex;
    justify-content: space-between;
    align-items: center;
    min-height: 50px;
    width: 100%;
    color: colors.$secondary-color;
    border: 1px solid colors.$secondary-color;
    padding: 10px 25px;
    border-radius: 5px;

    &.checked {
        opacity:0.4;

        .index,
        .title {
            text-decoration: line-through;
        }
    }

    &:not(:last-of-type) {
        margin-bottom: 20px;
    }

    & >:not(:last-child) {
        margin-right: 10px;
    }

    input[type="checkbox"] {
        flex: 0 0 auto;
        width: 20px;
        height: 20px;
    }

    input[type="checkbox"]:hover {
        cursor: pointer;
    }

    .title {
        flex: 1 1 auto;
    }

    button {
        flex: 0 0 auto;
        width: 25px;
        height: 25px;
        font-size: 20px;
        line-height: 25px;
        display: flex;
        justify-content: center;
        align-items: center;
        background-color: colors.$secondary-color;
        color: colors.$primary-color;
        transition: all 0.2s ease;

        &:hover {
            background-color: colors.$helper-color;
        }
    }
}
