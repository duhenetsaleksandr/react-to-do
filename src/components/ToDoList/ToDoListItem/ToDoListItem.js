import React from 'react';
import './todo-list-item.scss';

const ToDoListItem = ({ todo, deleteTodo, onEdit, toggleTodo }) => {
    const { id, title, completed } = todo;
    const classes = ['todos__list__item'];
    if (completed) classes.push('checked');

    return (
        <li className={classes.join(' ')}>
            <input
                type="checkbox"
                checked={completed}
                onChange={toggleTodo.bind(null, id)}
            />
            <div className="title">{title}</div>
            <button className="edit" onClick={onEdit.bind(null, todo)}>&#9998;</button>
            <button className="remove" onClick={deleteTodo.bind(null, id)}>&times;</button>
        </li>
    );
}

export default ToDoListItem;
