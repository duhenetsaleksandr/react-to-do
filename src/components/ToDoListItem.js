import React, { useContext } from 'react';
import '../styles/todo-list-item.scss';
import RequestContext from '../context';

export default function ToDoListItem(props) {
    const { toggleTodo, deleteTodo } = useContext(RequestContext);
    const { id, title, completed } = props.todo;
    const classes = ['todos__list__item'];
    if (completed) classes.push('checked');
    return (
        <li className={classes.join(' ')}>
            <input
                type="checkbox"
                checked={completed}
                onChange={toggleTodo.bind(null, id)}
            />
            <div className="index">{props.index + 1}</div>
            <div className="title">{title}</div>
            <button className="edit" onClick={props.onEdit.bind(null, props.todo)}>&#9998;</button>
            <button className="remove" onClick={deleteTodo.bind(null, id)}>&times;</button>
        </li>
    );
}
