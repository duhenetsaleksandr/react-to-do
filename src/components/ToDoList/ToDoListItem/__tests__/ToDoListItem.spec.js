import React from 'react';
import { act } from '@testing-library/react';
import ToDoListItem from '../ToDoListItem';

const props = {
    todo: {
        id: 1,
        title: 'create app',
        completed: false
    },
    deleteTodo: jest.fn(),
    onEdit: jest.fn(),
    toggleTodo: jest.fn(),
};

const checkedTodo = {
    id: 1,
    title: 'create app',
    completed: true
};

describe('should render ToDoListItem component', () => {
    describe('with default props', () => {
        let component;
        beforeEach(() => {
            component = mount(<ToDoListItem {...props} />);
        });

        it('should contain tag li', () => {
            const listItem = component.find('li');
            expect(listItem).toHaveLength(1);
            expect(listItem.getElement().props.className).not.toMatch(/checked/);
        });

        it('should contain input checkbox', () => {
            const input = component.find('input[type="checkbox"]');
            expect(input).toHaveLength(1);
        });

        it('should contain title', () => {
            const title = component.find('div.title');
            expect(title).toHaveLength(1);
            expect(title.getElement().props.children).toBe(props.todo.title);
        });

        it('should contain two buttons', () => {
            const button = component.find('button');
            expect(button).toHaveLength(2);
            expect(button.at(0).getElement().props.className).toBe('edit');
            expect(button.at(1).getElement().props.className).toBe('remove');
        });

        it('should call toggleTodo if input checkbox change', () => {
            const input = component.find('input[type="checkbox"]');
            act(()=> {
                input.getElement().props.onChange();
            })
            expect(props.toggleTodo).toHaveBeenCalledWith(props.todo.id);
        });

        it('should call onEdit if edit button clicked', () => {
            const button = component.find('button.edit');
            act(()=> {
                button.getElement().props.onClick();
            })
            expect(props.onEdit).toHaveBeenCalledWith(props.todo);
        });

        it('should call deleteTodo if remove button clicked', () => {
            const button = component.find('button.remove');
            act(()=> {
                button.getElement().props.onClick();
            })
            expect(props.deleteTodo).toHaveBeenCalledWith(props.todo.id);
        });

        it('should render ToDoListItem component snapshot', () => {
            const component = shallow(<ToDoListItem {...props} />);
            expect(component).toMatchSnapshot();
        });
    });

    describe('with custom props', () => {
        it('should li contain class checked if todo is completed', () => {
            const component = mount(<ToDoListItem {...props} todo={checkedTodo} />);
            const listItem = component.find('li');
            expect(listItem.getElement().props.className).toMatch(/checked/);
        });
    });
});
