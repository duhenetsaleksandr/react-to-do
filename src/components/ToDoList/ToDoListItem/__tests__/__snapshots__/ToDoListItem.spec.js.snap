// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ToDoListItem component with default props should render ToDoListItem component snapshot 1`] = `
<li
  className="todos__list__item"
>
  <input
    checked={false}
    onChange={[Function]}
    type="checkbox"
  />
  <div
    className="title"
  >
    create app
  </div>
  <button
    className="edit"
    onClick={[Function]}
  >
    ✎
  </button>
  <button
    className="remove"
    onClick={[Function]}
  >
    ×
  </button>
</li>
`;
